app/HutchAgent/Services/WorkflowFetchService.cs

```diff
@@ -0,0 +1,136 @@
+    {
+      var clientStream = await client.GetStreamAsync(downloadAddress) ??
+                         throw new InvalidOperationException("Invalid download URI");
+
+      await using var file =
+        File.OpenWrite(Path.Combine(workflowJob.WorkingDirectory.BagItPayloadPath(), "workflows.zip"));
+      await clientStream.CopyToAsync(file);
+      _logger.LogInformation("Successfully downloaded workflow from Workflow Hub.");
+    }
+
+    var workflowCrateExtractPath =
+      Path.Combine(workflowJob.WorkingDirectory.BagItPayloadPath(), "workflow", workflowId);
+    using (var archive =
+           new ZipArchive(File.OpenRead(Path.Combine(workflowJob.WorkingDirectory.BagItPayloadPath(),
+             "workflows.zip"))))
+    {
+      Directory.CreateDirectory(workflowCrateExtractPath);
+      archive.ExtractToDirectory(workflowCrateExtractPath);
+      _logger.LogInformation(
+        $"Unpacked workflow to {Path.Combine(workflowJob.WorkingDirectory.BagItPayloadPath(), "workflow", workflowId)}");
+    }
+
+    // Validate the Crate
+    try
+    {
+      // Validate that it's a crate at all, by trying to Initialise it
+      _crates.InitialiseCrate(workflowCrateExtractPath);
+    }
+    catch (Exception e) when (e is CrateReadException || e is MetadataException)
+    {
+      _logger.LogError(message: "RO-Crate downloaded is not a valid RO-Crate");
+      // Set ActionStatus to Failed and save updated
+      downloadAction.SetProperty("endTime", DateTime.Now);
+      _crates.UpdateCrateActionStatus(ActionStatus.FailedActionStatus, downloadAction);
+      roCrate.Save(workflowJob.WorkingDirectory.BagItPayloadPath());
+      throw;
+    }
+
+    //Set DownloadAction status to Completed
+    _crates.UpdateCrateActionStatus(ActionStatus.CompletedActionStatus, downloadAction);
+
+    downloadAction.SetProperty("endTime", DateTime.Now);
+    downloadAction.SetProperty("result", new Part()
+    {
+      Id = Path.Combine("workflow", workflowId)
+    });
+
+    var workflowEntity = new Entity(roCrate);
+    workflowEntity.SetProperty("@id", Path.Combine("workflow", workflowId));
+    var property = roCrate.Entities[mainEntity.Id];
+    workflowEntity.SetProperty("sameAs", new Part()
+    {
+      Id = property.Id
+    });
+    workflowEntity.SetProperty("@type", property.Properties["@type"]);
+    workflowEntity.SetProperty("name", property.Properties["name"]);
+    workflowEntity.SetProperty("conformsTo", property.Properties["conformsTo"]);
+    workflowEntity.SetProperty("distribution", property.Properties["distribution"]);
+    roCrate.Add(workflowEntity);
+    roCrate.RootDataset.AppendTo("mentions", downloadAction);
+    roCrate.Save(workflowJob.WorkingDirectory.BagItPayloadPath());
+    _logger.LogInformation($"Saved updated RO-Crate to {workflowJob.WorkingDirectory.BagItPayloadPath()}.");
+
+    return roCrate;
+  }
+
+  private static Entity CreateDownloadAction(ROCrate roCrate, string downloadAddress)
```

Comment:
Could update the return type to be ContextEntity since that's the exact type you're returning